apis/v1alpha2/shared_types.go

```diff
@@ -383,3 +383,54 @@ type AnnotationKey string
 // +kubebuilder:validation:MinLength=0
 // +kubebuilder:validation:MaxLength=4096
 type AnnotationValue string
+
+// TrafficMatches defines AddressMatch rules for inbound traffic according to
+// source and/or destination of that traffic.
+type TrafficMatches struct {
+	// SourceAddresses indicates the originating (source) network
+	// addresses which are valid for routing traffic.
+	//
+	// Support: Core
+	SourceAddresses []AddressMatch `json:"sourceAddresses"`
+
+	// DestinationAddresses indicates the destination network addresses
+	// which are valid for routing traffic.
+	//
+	// Support: Core
+	DestinationAddresses []AddressMatch `json:"destinationAddresses"`
+}
+
+// AddressMatch defines matching rules for network addresses by type.
+type AddressMatch struct {
+	// Type of the address, either IPAddress or NamedAddress.
+	//
+	// If NamedAddress is used this is a custom and specific value for each
+	// implementation to handle (and add validation for) according to their
+	// own needs.
+	//
+	// For IPAddress the implementor may expect either IPv4 or IPv6.
+	//
+	// Support: Core (IPAddress)
+	// Support: Custom (NamedAddress)
+	//
+	// +optional
+	// +kubebuilder:validation:Enum=IPAddress;NamedAddress
+	// +kubebuilder:default=IPAddress
+	Type *AddressType `json:"type,omitempty"`
+
+	// Value of the address. The validity of the values will depend
+	// on the type and support by the controller.
+	//
+	// If implementations support proxy-protocol (see:
+	// https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt) they
+	// must respect the connection metadata from proxy-protocol
+	// in the match logic implemented for these address values.
+	//
+	// Examples: `1.2.3.4`, `128::1`, `my-named-address`.
+	//
```

Comment:
Can we be explicit if CIDRs are allowed or not?

Reply:
It's inherent to the existing `IPAddressType` type, but yes I added documentation to that type explicitly pointing out that CIDR ranges are not allowed:

a6a982a7901da297494387d4ef62ebd880e48d7a

Reply:
IPAddressType is not used it seems. I think that is because you introduced L4AddressType to fix the problem with CRD generation.

Reply:
I think this is still accurate as is, the reference to `IPAddress` as an enum value here references the actual string `"IPAddress"` which is in turn type referenced by the `IPAddressType`? :thinking:

Reply:
Added a comment below, I think we also need type definitions for the new `AddressType`. Maybe both this and `GatewayAddressType` should be moved into a new `address_types.go` file?

Reply:
> It's inherent to the existing `IPAddressType` type,

I'm not sure about this. There's no explicit validation, but do we document that you are allowed to put a CIDR range in `GatewayAddress`?

Reply:
I see .. that makes sense. Thanks!

Reply:
> Added a comment below, I think we also need type definitions for the new AddressType. Maybe both this and GatewayAddressType should be moved into a new address_types.go file?

@robscott I might be misunderstanding you, but at least currently (now named `IPAddressType`) I don't think we need additional types, as `IPAddress` and `NamedAddress` are in use by way of the enum? :thinking:

Reply:
Sorry for the confusion here, I'm really just suggesting one of two things:

1. We just use AddressType and move the enum validation from the type definition to each field definition.
2. We split AddressType as you've currently proposed, and if so, we define valid values for the new type. For example, with the current AddressType, we have:

```go
type AddressType string

const (
    IPAddressType AddressType = "IPAddress"
    HostnameAddressType AddressType = "Hostname"
    // ...
)
```

But with the new type you've added, we only have the type definition, not the corresponding consts. The more I think about it though, the more I'm leaning towards just moving the enums to the fields instead of having a separate type everywhere we need an address in the API.

Reply:
I think @jpeach is also suggesting something similar to 1 in his [comment below](https://github.com/kubernetes-sigs/gateway-api/pull/976#discussion_r779965079). In both cases we're suggesting a shared type, the only distinction would be how we handle the difference in enum validation depending on where it's used.

Reply:
As suggested I've re-converged the types. I particularly made `AddressType` a shared type as this seems like the type we want to center on.

Everything looks right _except_ we're back to the problem where the generators are adding `Hostname` to the L4 matching address type. I've tried to override or cancel this but I'm not sure if that can be done with the generator tags. If someone has a hot tip on how this can be done I need some help.

Reply:
This has been changed as suggested, there is now only one `AddressType` :+1: